### DISCUSSION TITLE
A new article about TeaVM

### DISCUSSION THREAD

**Discussion by konsoletyper** [konsoletyper/teavm#875] [Announcements]
This article compares TeaVM and Kotlin/JS performance using as example a software 3D renderer: https://teavm.hashnode.dev/comparing-teavm-with-kotlinjs-peformance-with-software-3d-renderer

**Comment by bertung** [comment]
Interesting article Alex, I have a mac pro 2013, 12 Core, 64GB ram, AMD Fire pro dedicated GPU . My results show on all modern browsers I tested, Kotlin produced code is always 2x slower than teavm/JS and teavm/wasm. On chrome, teavm/wasm is the clear winner. I will wager that it may be Graphics card related. Graphics based benchmarking will always prove wrong. You do not mention your graphics card on your article. It should be the culprit.

**Reply by konsoletyper** [reply]
@bertung, I don't think graphics card is involved here. The 3D renderer, presented in the article, utilizes only CPU. That was the goal: to write a *software* 3D renderer to compare performance of different compilers.

**Reply by bertung** [reply]
Well, you still utilize GPU to write that chunk of data to the graphics card to display something even though you do not use WebGL.

**Reply by konsoletyper** [reply]
@bertung you are right, but I only measure that part of calculations that performs rasterization in main memory. Further calculations related to passing buffers between main script and worker fall into other metric. So you should not take into account "Average frame rendering time, microseconds". Instead, look into "Average by worker" metric, which is a primary metric used in "Benchmark" section in the article.
